test/hotspot/jtreg/compiler/c2/irTests/UDivINodeIdealizationTests.java

```diff
@@ -0,0 +1,148 @@
+            Asserts.assertFalse(shouldThrow, "Expected an exception to be thrown.");
+        } catch (ArithmeticException e) {
+            Asserts.assertTrue(shouldThrow, "Did not expect an exception to be thrown.");
+        }
+
+        try {
+            Asserts.assertEQ(Integer.divideUnsigned((a * b), b), retainDenominator(a, b));
+            Asserts.assertFalse(shouldThrow, "Expected an exception to be thrown.");
+        } catch (ArithmeticException e) {
+            Asserts.assertTrue(shouldThrow, "Did not expect an exception to be thrown.");
+        }
+
+        Asserts.assertEQ(Integer.divideUnsigned(a, 1), identity(a));
+        Asserts.assertEQ(Integer.divideUnsigned(a, Integer.divideUnsigned(13, 13)), identityAgain(a));
+        Asserts.assertEQ(Integer.divideUnsigned(a, 8), divByPow2(a), "divByPow2 " + a);
+        Asserts.assertEQ(Integer.divideUnsigned(a, -2147483648), divByPow2Big(a));
+        Asserts.assertEQ(Integer.divideUnsigned(a, Integer.divideUnsigned(-2129457054, -2129457054)), identityAgainButBig(a));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    @IR(counts = {IRNode.DIV_BY_ZERO_TRAP, "1"})
+    // Checks x / x => 1
+    public int constant(int x) {
+        return Integer.divideUnsigned(x, x);
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    // Checks x / 1 => x
+    public int identity(int x) {
+        return Integer.divideUnsigned(x, 1);
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    // Checks x / (c / c) => x
+    public int identityAgain(int x) {
+        return Integer.divideUnsigned(x, Integer.divideUnsigned(13, 13));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    // Checks x / (c / c) => x
+    public int identityAgainButBig(int x) {
+        return Integer.divideUnsigned(x, Integer.divideUnsigned(-2129457054, -2129457054));  // Integer.parseUnsignedInt("2165510242") = -2129457054
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    @IR(counts = {IRNode.DIV_BY_ZERO_TRAP, "1"})
+    // Checks x / (y / y) => x
+    public int identityThird(int x, int y) {
+        return Integer.divideUnsigned(x, Integer.divideUnsigned(y, y));
+    }
+
+    @Test
+    @IR(counts = {IRNode.MUL, "1",
+            IRNode.UDIV_I, "1",
+            IRNode.DIV_BY_ZERO_TRAP, "1"
+    })
+    // Hotspot should keep the division because it may cause a division by zero trap
+    public int retainDenominator(int x, int y) {
+        return Integer.divideUnsigned((x * y), y);
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    @IR(counts = {IRNode.URSHIFT, "1"})
+    // Dividing an unsigned number by 8 is a trivial right shift by 3
+    public int divByPow2(int x) {
+        return Integer.divideUnsigned(x, 8);
+    }
+
+    @Test
+    @IR(failOn = {IRNode.DIV})
+    @IR(counts = {IRNode.URSHIFT, "1"})
+    public int divByPow2Big(int x) {
+        return Integer.divideUnsigned(x, -2147483648);  // -2147483648 = Integer.parseUnsignedInt("2147483648")
```

Comment:
You should use `Integer.MIN_VALUE`

Reply:
In general, I think `1 << someValue` would be better than the decimal representation.